core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinMemoizedRecordCursorFactory.java

```diff
@@ -0,0 +1,355 @@
+            CairoConfiguration configuration,
+            RecordMetadata metadata,
+            RecordCursorFactory masterFactory,
+            RecordCursorFactory slaveFactory,
+            int columnSplit,
+            int slaveSymbolColumnIndex,
+            AsofJoinColumnAccessHelper columnAccessHelper,
+            JoinContext joinContext,
+            long toleranceInterval
+    ) {
+        super(metadata, joinContext, masterFactory, slaveFactory);
+        assert slaveFactory.supportsTimeFrameCursor();
+        this.columnAccessHelper = columnAccessHelper;
+        this.toleranceInterval = toleranceInterval;
+        this.slaveSymbolColumnIndex = slaveSymbolColumnIndex;
+        this.cursor = new AsOfJoinMemoizedRecordCursor(
+                columnSplit,
+                NullRecordFactory.getInstance(slaveFactory.getMetadata()),
+                masterFactory.getMetadata().getTimestampIndex(),
+                masterFactory.getMetadata().getTimestampType(),
+                slaveFactory.getMetadata().getTimestampIndex(),
+                slaveFactory.getMetadata().getTimestampType(),
+                configuration.getSqlAsOfJoinLookAhead()
+        );
+    }
+
+    @Override
+    public boolean followedOrderByAdvice() {
+        return masterFactory.followedOrderByAdvice();
+    }
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) throws SqlException {
+        RecordCursor masterCursor = masterFactory.getCursor(executionContext);
+        TimeFrameRecordCursor slaveCursor = null;
+        try {
+            slaveCursor = slaveFactory.getTimeFrameCursor(executionContext);
+            cursor.of(masterCursor, slaveCursor, executionContext.getCircuitBreaker());
+            return cursor;
+        } catch (Throwable e) {
+            Misc.free(slaveCursor);
+            Misc.free(masterCursor);
+            throw e;
+        }
+    }
+
+    @Override
+    public int getScanDirection() {
+        return masterFactory.getScanDirection();
+    }
+
+    @Override
+    public boolean recordCursorSupportsRandomAccess() {
+        return false;
+    }
+
+    @Override
+    public void toPlan(PlanSink sink) {
+        sink.type("AsOf Join Memoized Scan");
+        sink.attr("condition").val(joinContext);
+        sink.child(masterFactory);
+        sink.child(slaveFactory);
+    }
+
+    @Override
+    protected void _close() {
+        Misc.freeIfCloseable(getMetadata());
+        Misc.free(masterFactory);
+        Misc.free(slaveFactory);
+    }
+
+    private class AsOfJoinMemoizedRecordCursor extends AbstractKeyedAsOfJoinRecordCursor {
+
+        private static final long NOT_REMEMBERED = Long.MIN_VALUE;
+
+        // These three hashmaps are used in strict parallel, always putting the same key in each.
+        // When accessing them, we can look up the key index only once, and use it for all three maps.
+        private final IntLongHashMap symKeyToRowId = new IntLongHashMap();
```

Comment:
Having three hash tables means that we have 3x overhead for the keys. In case of high cardinality symbol column, the overhead would be noticeable. Also, the choice of the identity hash function in the hash table is not ideal considering that symbol keys are usually sequential integers - after modulo, the clusters are more likely with identity function. Consider switching to `OrderedMap`/`Unordered8Map` or implementing a new on-heap map.